File: lib/engine/note_block.cxx
/*
** note_block.cxx
** Login : <elthariel@rincevent>
** Started on  Sat Jul 31 22:26:04 2010 elthariel
** $Id$
**
** Author(s):
**  - elthariel <elthariel@gmail.com>
**
** Copyright (C) 2010 elthariel
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 3 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#include <iostream>
#include <cstring>
#include <assert.h>

#include "note_block.hh"
#include "memory.hh"

using namespace std;

NoteBlock::NoteBlock()
{
  reset();
}

NoteBlock::NoteBlock(const NoteBlock &a_block)
  : Block(a_block)
{
  reset();

  for (int i = 0; i < PPQ * 4 * MAX_BARS; ++i)
    if (a_block.data[i])
    {
      Event     *a_iter = a_block.data[i];
      Event     *iter = new Event(*a_iter);

      data[i] = iter;

      while(a_iter->next)
      {
        iter->next = new Event(*a_iter->next);

        iter = iter->next;
        a_iter = a_iter->next;
      }
    }
}

NoteBlock::~NoteBlock()
{
  for (int i = 0; i < PPQ * 4 * MAX_BARS; ++i)
    if (data[i])
    {
      Event     *iter = data[i];
      Event     *to_delete = 0;

      while (iter)
      {
        to_delete = iter;
        iter = iter->next;
        delete to_delete;
      }

      data[i] = 0;
    }
}

void          *NoteBlock::operator new(size_t sz)
{
  assert(sz == sizeof(NoteBlock));

  return Memory::note_block().alloc();
}

void          NoteBlock::operator delete(void *p)
{
  if (p)
    Memory::note_block().dealloc((NoteBlock *)p);
}

Event         *NoteBlock::operator[](unsigned int tick)
{
  assert (tick < PPQ * 4 * MAX_BARS);

  if (tick >= PPQ * 4 * length)
    tick = PPQ * 4 * length - 1;

  return data[tick];
}

Event         *NoteBlock::get_note_on_tick(unsigned int tick,
                                        unsigned int max_bar,
                                        unsigned char note)
{
  unsigned int i;
  Event *iter = 0;

  assert(max_bar < MAX_BARS);

  if (max_bar > length)
    max_bar = length;

  for (i = 0; i < max_bar * 4 * PPQ; i++)
  {
    iter = data[i];
    while (iter)
    {
      if (iter->midi[1] == note && tick >= i && tick < i + iter->duration)
        return iter;
      iter = iter->next;
    }
  }
  return iter;
}

bool          NoteBlock::insert(unsigned int tick, Event *e)
{
  assert(tick < MAX_BARS * PPQ * 4);
  assert(e == 0);

  if (tick >= MAX_BARS * PPQ * 4 || e == 0)
    return false;

  Event *iter = data[tick];

  e->next = data[tick];
  data[tick] = e;

  return true;
}

bool          NoteBlock::remove(unsigned int tick, Event *e)
{
  assert(tick < MAX_BARS * PPQ * 4);
  assert(e == 0);

  if (tick > MAX_BARS * PPQ * 4 || e == 0)
    return false;

  Event *iter = data[tick];

  if (iter == e)
  {
    data[tick] = iter->next;
    return true;
  }

  while (iter && iter->next)
  {
    if (iter->next == e)
    {
      iter->next = iter->next->next;
      return true;
    }
    iter = iter->next;
  }
  return false;
}

int           NoteBlock::next_used_tick(unsigned int tick)
{
  unsigned int i;

  for (i = tick; i < MAX_BARS * PPQ * 4; i++)
    if (data[i])
      return i;
  return -1;
}

void          NoteBlock::reset()
{
  Block::reset();
  memset(&data, 0, sizeof(data));
}

File: lib/controller/note_block.rb
##
## phrase.rb
## Login : <elthariel@rincevent>
## Started on  Sat Jul 17 18:50:12 2010 elthariel
## $Id$
##
## Author(s):
##  - elthariel <elthariel@gmail.com>
##
## Copyright (C) 2010 elthariel
## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
##

class NoteBlockController
  attr_reader :block

  def initialize(controller, block)
    @controller = controller
    @roll = controller.pianoroll
    @block = block
    # @phrase = @pattern.get(track_id)
  end

  def alloc_event!
    e = Kiara::Memory.event.alloc
    e.reset!
    e
  end

  def dealloc_event!(e)
    if e
      e.reset!
      Kiara::Memory.event.dealloc(e)
    end
  end

  # Pos has the same syntax than cursor
  def occupied?(pos)
    @block.get_note_on_tick pos[0], @block.length, pos[1]
  end

  def get_note(pos)
    iter = @block.get(pos[0])
    while iter and iter.data1 != pos[1] do
      iter = iter.next
    end
    return iter if iter and iter.noteon?
    nil
  end

  def insert!(tick, event)
    return true if event and @block.insert!(tick, event)
    false
  end

  # pos = [tick, note]
  # returns true if the note has been moved at the request pos
  def move_note!(old_pos, new_pos)
    return true if old_pos == new_pos
    if old_pos[0] == new_pos[0] and (note = get_note old_pos)
      note.data1 = new_pos[1]
    elsif (note = delete_note_on_tick! old_pos[0], old_pos[1])
      note.data1 = new_pos[1]
      unless insert!(new_pos[0], note)
        dealloc_event note
        return false
      end
    else
      return false
    end
    true
  end

  # Resize the note at pos, by offset tick
  # returns true if successfull
  def resize_note!(pos, offset)
    if (note = get_note(pos))
      if note.duration + offset < 1
        @controller.pianoroll.note_duration = 1
        note.duration = 1
      else
        @controller.pianoroll.note_duration = note.duration + offset
        note.duration = note.duration + offset
      end
      true
    else
      false
    end
  end


  # Delete a note with data1=note starting precisely at tick=tick
  # Returns true if correctly deleted
  def delete_note_on_tick!(tick, note)
    iter = @block.get(tick)
    while iter and iter.data1 != note do
      iter = iter.next
    end
    return iter if iter and @block.remove!(tick, iter)
    nil
  end

  # Delete a note with data1=note starting at or overlapping tick=tick
  # Return true if the event was deleted
  def delete_note_on_tick_overlapping!(pos)
    event = occupied?(pos)
    #  if event
    if event and event.noteon?
      tick = nil
      each_pos do |t, e|
        if pos[0] >= t and pos[0] < t + event.duration and pos[1] == event.data1
          tick = t
        end
      end
      return event if tick and @block.remove!(tick, event)
      nil
    end
    nil
  end

  # phrase.each_pos { |event| ... } Iterates on all events of
  # the phrase, passing each event to the given block
  def each
    iter = 0
    while (iter = @block.next_used_tick iter) >= 0 do
      e = @block.get(iter)
      while e do
        yield e
        e = e.next
      end
      iter += 1
    end
  end

  # phrase.each_pos { |tick, event| ... } Iterates on all events of
  # the phrase, passing the tick of the current event and the event to
  # the block
  def each_pos
    iter = 0
    while (iter = @block.next_used_tick iter) >= 0 do
      e = @phrase.get(iter)
      while e do
        yield iter, e
        e = e.next
      end
      iter += 1
    end
  end

end

